documentation/spec-opt.md

```diff
@@ -0,0 +1,7 @@
+# Speculative Optimization
+
+Not all RIR code is converted into PIR, only code which is re-used a lot (currently, when a function is called 3 times it gets optimized). This is because, the PIR optimizations themselves take a long time (remember, this is during interpretation) - if an instruction is only run once, it's faster to just run it unoptimized. Furthermore, PIR performs **speculative optimizations**, and it could use the results of previous iterations to formi its assumptions.
```

Comment:
I think there's now a `RIR_WARMUP` or `PIR_WARMUP` flag that sets the threshold? So it's not necessarily 3.